tarsier/tag_utils.ts

```diff
@@ -458,67 +461,124 @@ function insertTags(
+  function createAndInsertTag(
+    el: HTMLElement,
+    xpath: string,
+    textNodeIndex: number | null,
+    isAbsolutelyPositioned: boolean,
+    referenceNode: ChildNode | null = null,
+    originalTextContent: string | null = null,
+  ) {
+    const symbol = getTagSymbol(el);
+    const idSpan = create_tagged_span(idNum, symbol, el);
+
+    tagDataList.push({
+      xpath,
+      element: el,
+      tagElement: idSpan,
+      textNodeIndex,
+      originalTextContent,
+    });
+
+    if (referenceNode && el.parentElement) {
+      el.insertBefore(idSpan, referenceNode);
+    } else if (isTextInsertable(el) && el.parentElement) {
+      el.parentElement.insertBefore(idSpan, el);
+    } else {
+      const insertionElement = getElementToInsertInto(el);
+      insertionElement.prepend(idSpan);
+      if (isAbsolutelyPositioned) {
         absolutelyPositionTagIfMisaligned(idSpan, insertionElement);
       }
-      idNum++;
-    } else if (isImageElement(el)) {
-      // Handle image elements
-      const idSpan = create_tagged_span(idNum, el);
-      if (el.parentElement) {
-        el.parentElement.insertBefore(idSpan, el);
-        absolutelyPositionTagIfMisaligned(idSpan, el);
-      }
-      idNum++;
+    }
+
+    if (isAbsolutelyPositioned && !referenceNode) {
+      absolutelyPositionTagIfMisaligned(idSpan, el);
+    }
+
+    idNum++;
+  }
+
+  const tagDataList: {
+    xpath: string;
+    element: HTMLElement;
+    tagElement: HTMLElement;
+    textNodeIndex: number | null;
+    originalTextContent: string | null;
+  }[] = [];
+  let idNum = 0;
+
+  for (const el of elementsToTag) {
+    const xpath = getElementXPath(el);
+
+    if (isInteractable(el) || isImageElement(el)) {
+      const isAbsolutelyPositioned =
+        !isTextInsertable(el) || isImageElement(el);
+
+      const originalTextContent = el.textContent?.trim() || null;
+
+      createAndInsertTag(
+        el,
+        xpath,
+        null,
+        isAbsolutelyPositioned,
+        null,
+        originalTextContent,
+      );
     } else if (tagLeafTexts) {
       trimTextNodeStart(el);
-      const validTextNodes = Array.from(el.childNodes).filter(
-        isTaggableTextNode,
-      );
-      const allTextNodes = Array.from(el.childNodes).filter(
+      const textNodes = Array.from(el.childNodes).filter(
         (child) => child.nodeType === Node.TEXT_NODE,
       );
-      for (let child of validTextNodes) {
-        const parentXPath = getElementXPath(el);
-        const textNodeIndex = allTextNodes.indexOf(child) + 1;
-
-        idToTagMetadata[idNum] = {
-          xpath: parentXPath,
-          textNodeIndex: textNodeIndex,
-        };
-
-        const idSpan = create_tagged_span(idNum, el);
-        el.insertBefore(idSpan, child);
-        idNum++;
-      }
+      const validTextNodes = textNodes.filter(isTaggableTextNode);
+
+      validTextNodes.forEach((child) => {
+        const textNodeIndex = textNodes.indexOf(child) + 1;
+
```

Comment:
we should add a test where we have invalid text nodes within the html and ensure we are getting the correct index